File: dataStruct_2021/Tree/BinaryTree.m
//
//  BinaryTree.m
//  dataStruct
//
//  Created by 张森 on 2021/4/21.
//

#import "BinaryTree.h"
#import "Queue.h"
#import "MJBinaryTreeInfo.h"

@implementation BinaryTreeNode
+ (instancetype)treeNode:(id<Compare>)node {
    
    BinaryTreeNode *treeNode = [[self alloc] init];
    treeNode.element = node;
    return treeNode;
}

- (BOOL)isLeaf {
    
    return !self.left && !self.right;
}

- (BOOL)hasTwoChild {
    
    return self.left && self.right;
}

/// 是否是父节点的左子树
- (BOOL)isLeftChild {
    
    return self.parent.left == self;
}
/// 是否是父节点的右子树
- (BOOL)isRightChild {
    
    return self.parent.right == self;
}

/// 返回兄弟节点
- (instancetype)brotherNode {
    
    if ([self isLeftChild]) {
        return self.parent.right;
    } else if ([self isRightChild]) {
        return self.parent.left;
    } else {
        return nil;
    }
}

@end
@interface BinaryTree ()<MJBinaryTreeInfo>

@end
@implementation BinaryTree

#pragma mark - 基本操作
+ (instancetype)treeWithCompartor:(nullable TreeCompartor *)compartor {
    
    BinaryTree *tree = [[self alloc] init];
    tree.compartor = compartor;
    return tree;
}

- (NSInteger)size {
    
    return _size;
}

/// 计算树的高度
- (NSInteger)height {
    
    Queue *queue = [Queue queue];
    [queue push:self.rootNode];
    NSInteger height = 0;
    // 每一层的数量
    NSInteger levelSize = 1;
    while ([queue size] > 0) {
        
        BinaryTreeNode *node = [queue front];
        if (node.left) {
            
            [queue push:node.left];
        }
        
        if (node.right) {
            
            [queue push:node.right];
        }
        [queue pop];
        // 每遍历完一个元素,当前层的元素数量就减一
        levelSize --;
        if (levelSize == 0) {
            // 当前层遍历完后,队列的数量就是下一层元素的数量
            levelSize = [queue size];
            // 遍历完一层,高度加一
            height ++;
        }
    }
    return height;
    // 递归方法
//    return [self _height:self.rootNode];
}

//- (NSInteger)_height:(BinaryTreeNode *)node {
//
//    if (!node) {
//        return 0;
//    }
//    return 1 + MAX([self _height:node.left], [self _height:node.right]);
//}

/// 是否是完全二叉树
- (BOOL)isCompleteTree {
    
    if (!self.rootNode) {
        return NO;
    }
    
    // 1. 采用层序遍历的方式来进行判断
    // 2. 如果一个节点只有左子树,没有右子树;或两边都没有时,证明从这个节点之后的节点都是叶子节点
    //      - 在后面的节点中,应该都是叶子节点,即左右子树都不存在,否则,返回NO
    // 3. 如果有右子树,没有左子树,直接返回NO
    Queue *queue = [Queue queue];
    [queue push:self.rootNode];
    
    BOOL leaf = NO;
    while ([queue size] > 0) {
        
        BinaryTreeNode *node = [queue front];
        [queue pop];
        
        if (leaf && ![node isLeaf]) {
            // 如果是叶子节点,且有左右子树,直接返回NO
            return NO;
        }
        
        if (node.left) {
            // 有左边,则入队
            [queue push:node.left];
        } else if (node.right) {
            // !node.left
            // 没有左边,且有右边,则直接返回NO
            return NO;
        }
        
        if (node.right) {
            // 有右边,则入队
            [queue push:node.right];
        } else {
            // 没有右节点,则证明是叶子节点
            leaf = YES;
        }
    }
    return YES;
}

- (BOOL)isEmpty {
    
    if (self.rootNode) {
        
        return YES;
    }
    return NO;
}

- (void)clear {
    
    self.rootNode = nil;
    self.size = 0;
}

#pragma mark - 遍历相关

static BOOL _preorderStop = NO;
/// 前序打印
- (void)preorderPrint:(void(^)(id node, BOOL *isStop))element {
    
    if (!element) {
        return;
    }
    _preorderStop = NO;
    [self _preorderPrint:self.rootNode element:element];
}

static BOOL _inorderStop = NO;
/// 中序打印
- (void)inorderPrint:(void(^)(id node, BOOL *isStop))element {
    
    if (!element) {
        return;
    }
    _inorderStop = NO;
    [self _inorderPrint:self.rootNode element:element];
}

static BOOL _postorderStop = NO;
/// 后序打印
- (void)postorderPrint:(void(^)(id node, BOOL *isStop))element {
    
    if (!element) {
        return;
    }
    _postorderStop = NO;
    [self _postorderPrint:self.rootNode element:element];
}

static BOOL _levelorderStop = NO;
/// 层序打印
- (void)levelorderPrint:(void(^)(id node, BOOL *isStop))element {
    
    _levelorderStop = NO;
    if (!self.rootNode || !element) {
        return;
    }
    Queue *queue = [Queue queue];
    [queue push:self.rootNode];
    while ([queue size] > 0) {
        
        BinaryTreeNode *node = [queue front];
        if (_levelorderStop) {
            _levelorderStop = NO;
            return;
        }
        element(node.element, &_levelorderStop);
        
        if (node.left) {
            
            [queue push:node.left];
        }
        
        if (node.right) {
            
            [queue push:node.right];
        }
        [queue pop];
    }
}

/// 私有方法前序打印
/// @param rootNode 根节点
/// @param element 遍历block
- (void)_preorderPrint:(BinaryTreeNode *)rootNode element:(void(^)(id node, BOOL *isStop))element {
    
    if (!rootNode || _preorderStop) {
        return;
    }
    
    if (!_preorderStop) {
        
        element(rootNode.element, &_preorderStop);
    }
    [self _preorderPrint:rootNode.left element:element];
    [self _preorderPrint:rootNode.right element:element];
}

/// 私有方法中序打印
/// @param rootNode 根节点
/// @param element 遍历block
- (void)_inorderPrint:(BinaryTreeNode *)rootNode element:(void(^)(id node, BOOL *isStop))element {
    
    if (!rootNode || _inorderStop) {
        return;
    }
    [self _inorderPrint:rootNode.left element:element];
    if (!_inorderStop) {
        
        element(rootNode.element, &_inorderStop);
    }
    [self _inorderPrint:rootNode.right element:element];
}

/// 私有方法后序打印
/// @param rootNode 根节点
/// @param element 遍历block
- (void)_postorderPrint:(BinaryTreeNode *)rootNode element:(void(^)(id node, BOOL *isStop))element {
    
    if (!rootNode || _postorderStop) {
        return;
    }
    [self _postorderPrint:rootNode.left element:element];
    [self _postorderPrint:rootNode.right element:element];
    if (!_postorderStop) {
        
        element(rootNode.element, &_postorderStop);
    }
}

- (NSInteger)_compare:(id<Compare>)element1 to:(id<Compare>)element2 {
    
    if (self.compartor) {
        return [self.compartor compare:element1 to:element2];
    }
    return [element1 compareTo:element2];
}

#pragma mark - 前驱后继

- (BinaryTreeNode *)_predecessor:(BinaryTreeNode *)node {
    // 前驱 -> 中序遍历中当前节点的前一个节点
    // 利用二叉搜索树来理解,即为找到比当前节点小的最大的数
    // 1. 如果有左子树,则从左子树中寻找最右下角的节点
    // 2. 如果没有左子树,则一直找parent,直到parent的右节点等于当前节点
    if (!node) {
        return nil;
    }
    if (node.left) {
        // 左子树不为空,则从左子树中一直往右找
        BinaryTreeNode *tempNode = node.left;
        while (tempNode.right) {
            tempNode = tempNode.right;
        }
        return tempNode;
    } else {
        // 左子树为空,需要往上找parent
        while (node.parent && node.parent.left == node) {
            node = node.parent;
        }
        // 退出循环
        // 1. node没有parent了,则证明没有前驱节点
        // 2. node为node.parent的右节点,则前驱为node.parent
        return node.parent;
    }
}

- (BinaryTreeNode *)_successor:(BinaryTreeNode *)node {
    // 后继 -> 中序遍历中当前节点的后一个节点
    // 利用二叉搜索树来理解,即为找到比当前节点大的最小的数
    // 1. 如果有右子树,则从右子树中寻找最左下角的节点
    // 2. 如果没有右子树,则一直找parent,直到parent的左节点等于当前节点
    if (!node) {
        return nil;
    }
    if (node.right) {
        // 右子树不为空,则从右子树中一直往左找
        BinaryTreeNode *tempNode = node.right;
        while (tempNode.left) {
            tempNode = tempNode.left;
        }
        return tempNode;
    } else {
        // 右子树为空,需要往上找parent
        while (node.parent && node.parent.right == node) {
            node = node.parent;
        }
        // 退出循环
        // 1. node没有parent了,则证明没有前驱节点
        // 2. node为node.parent的左节点,则前驱为node.parent
        return node.parent;
    }
}

#pragma mark - protpcol
- (id)root {
    return self.rootNode;
}

- (id)left:(BinaryTreeNode *)node {
    
    return node.left;
}

- (id)right:(BinaryTreeNode *)node {
    
    return node.right;
}

- (id)string:(BinaryTreeNode *)node {
    
    return [NSString stringWithFormat:@"%@(p-%@)", node.element, node.parent.element];
}

@end

File: dataStruct_2023/Tree/MJBinaryTrees/BinaryTree.swift
//
//  BinaryTree.swift
//  dataStruct_2023
//
//  Created by firstzhang on 2023/7/28.
//

import Foundation
class BinaryTreeNode<E>: CustomStringConvertible {
    var element: E
    var left: BinaryTreeNode?
    var right: BinaryTreeNode?
    var parent: BinaryTreeNode?
    init(element: E, left: BinaryTreeNode? = nil, right: BinaryTreeNode? = nil, parent: BinaryTreeNode? = nil) {
        self.element = element
        self.left = left
        self.right = right
        self.parent = parent
    }
    
    func isLeftChild() -> Bool {
        if parent == nil {
            return false
        }
        return parent!.left === self
    }
    
    func isRightChild() -> Bool {
        if parent == nil {
            return false
        }
        return parent!.right === self
    }
    
    deinit {
        print("BinaryTreeNode - \(element) - deinit")
    }
    
    var description: String {
        return "\(self.element)"
    }
}

class BinaryTree<E: Comparable>: NSObject, MJBinaryTreeInfo {
    var _size: Int = 0
    var rootNode: BinaryTreeNode<E>?
    var comparator: ((_ element1: E, _ element2: E) -> Int)?
    
    init(comparator: @escaping (_: any Comparable, _: any Comparable) -> Int) {
        self.comparator = comparator
    }
    
    override init() {
        super.init()
    }
    
    func size() -> Int {
        return _size
    }
    
    func add(element: E) {
        if self.checkIsNull(element: element) {
            return
        }
        let addNode = BinaryTreeNode(element: element)
        add(element: addNode)
    }
    
    func add(element: BinaryTreeNode<E>) {
        
    }
    
    func remove(element: E) -> E? {
        return nil
    }
    
    func contains(element: E) -> Bool {
        return node(element: element) != nil
    }
    
    func checkIsNull(element: E?) -> Bool {
        if element == nil {
            return true
        }
        return false
    }
    
    func node(element: E) -> BinaryTreeNode<E>? {
        return nil
    }
    
    /// 是否是叶子节点
    static func isLeaf(node: BinaryTreeNode<E>?) -> Bool {
        if node == nil {
            return false
        }
        return node!.left == nil && node!.right == nil
    }
    
    /// 是否是度为1的节点
    static func hasOnlyOneChild(node: BinaryTreeNode<E>?) -> Bool {
        if node == nil {
            return false
        }
        var childCount = 0
        if node!.left != nil {
            childCount += 1
        }
        if node!.right != nil {
            childCount += 1
        }
        return childCount == 1
    }
    
    /// 是否是度为2的节点
    static func hasTwoChild(node: BinaryTreeNode<E>?) -> Bool {
        if node == nil {
            return false
        }
        return node!.left != nil && node!.right != nil
    }
    
    // MARK: - Iterator
    func preorder(visitor: ((_ element: E) -> Void)?) {
        // 前序遍历
        preorder(node: rootNode, visitor: visitor)
    }

    private func preorder(node: BinaryTreeNode<E>?, visitor: ((_ element: E) -> Void)?) {
        if node == nil {
            return
        }
        if visitor != nil {
            visitor!(node!.element)
        }
        preorder(node: node!.left, visitor: visitor)
        preorder(node: node!.right, visitor: visitor)
    }
    
    func inorder(visitor: ((_ element: E) -> Void)?) {
        // 中序遍历
        inorder(node: rootNode, visitor: visitor)
    }
    
    private func inorder(node: BinaryTreeNode<E>?, visitor: ((_ element: E) -> Void)?) {
        if node == nil {
            return
        }
        inorder(node: node!.left, visitor: visitor)
        if visitor != nil {
            visitor!(node!.element)
        }
        inorder(node: node!.right, visitor: visitor)
    }
    
    func postorder(visitor: ((_ element: E) -> Void)?) {
        // 后序遍历
        postorder(node: rootNode, visitor: visitor)
    }
    
    private func postorder(node: BinaryTreeNode<E>?, visitor: ((_ element: E) -> Void)?) {
        if node == nil {
            return
        }
        postorder(node: node!.left, visitor: visitor)
        postorder(node: node!.right, visitor: visitor)
        if visitor != nil {
            visitor!(node!.element)
        }
    }
    
    func levelOrder(visitor: ((_ element: E) -> Void)?) {
        // 层序遍历
        if rootNode == nil {
            return
        }
        var queue = [BinaryTreeNode<E>]()
        queue.append(rootNode!)
        levelOrder(queue: queue, visitor: visitor)
    }
    func levelOrder(queue: Array<BinaryTreeNode<E>>, visitor: ((_ element: E) -> Void)?) {
        if queue.count == 0 {
            return
        }
        var tempQueue = queue
        let node = tempQueue.first
        if node!.left != nil {
            tempQueue.append(node!.left!)
        }
        if node!.right != nil {
            tempQueue.append(node!.right!)
        }
        if visitor != nil {
            visitor!(node!.element)
        }
        tempQueue.removeFirst()
        levelOrder(queue: tempQueue, visitor: visitor)
    }
    /// 前驱节点
    func precursor(node: BinaryTreeNode<E>?) -> BinaryTreeNode<E>? {
        if node == nil {
            return nil
        }
        // 当前节点有左子树
        // 当前节点的左子树中最右下角的节点
        if node!.left != nil {
            var precursorNode = node!.left!
            while precursorNode.right != nil {
                precursorNode = precursorNode.right!
            }
            return precursorNode
        }
        // 当前节点没有左子树
        // 找当前节点的parent,直到当前节点是parent的右子树
        // 当parent == nil时还不满足条件,则没有前驱节点
        var precursorNode = node!.parent
        var currentNode = node!
        while precursorNode != nil {
            if precursorNode!.right?.element == currentNode.element {
                return precursorNode
            }
            currentNode = precursorNode!
            precursorNode = precursorNode!.parent
        }
        return nil
    }
    /// 前驱节点
    func precursor(element: E) -> E? {
        let node = node(element: element)
        if node == nil {
            return nil
        }
        return precursor(node: node)?.element
    }
    
    /// 后继节点(只需要将前驱结点的代码left和right互换即可)
    func successor(node: BinaryTreeNode<E>?) -> BinaryTreeNode<E>? {
        if node == nil {
            return nil
        }
        // 当前节点有右子树
        // 当前节点的右子树中最左下角的节点
        if node!.right != nil {
            var successorNode = node!.right!
            while successorNode.left != nil {
                successorNode = successorNode.left!
            }
            return successorNode
        }
        // 当前节点没有右子树
        // 找当前节点的parent,直到当前节点是parent的左子树
        // 当parent == nil时还不满足条件,则没有前驱节点
        var successorNode = node!.parent
        var currentNode = node!
        while successorNode != nil {
            if successorNode!.left?.element == currentNode.element {
                return successorNode
            }
            currentNode = successorNode!
            successorNode = successorNode!.parent
        }
        return nil
    }
    
    /// 后继节点(只需要将前驱结点的代码left和right互换即可)
    func successor(element: E) -> E? {
        let node = node(element: element)
        if node == nil {
            return nil
        }
        return successor(node: node)?.element
    }
    // MARK: - MJBinaryTreeInfo
    func root() -> Any {
        return rootNode as Any
    }
    
    func left(_ node: Any) -> Any {
        let tempNode = node as? BinaryTreeNode<E>
        return tempNode?.left as Any
    }
    
    func right(_ node: Any) -> Any {
        let tempNode = node as? BinaryTreeNode<E>
        return tempNode?.right as Any
    }
    
    func string(_ node: Any) -> Any {
        let tempNode = node as? BinaryTreeNode<E>
        return tempNode?.description as Any
    }
}
